pennylane/interfaces/jacobian_products.py

```diff
@@ -189,32 +191,44 @@ class TransformJacobianProducts(JacobianProductCalculator):
     """
 
     def __repr__(self):
-        return f"TransformJacobianProducts({self._inner_execute}, gradient_transform={self._gradient_transform}, gradient_kwargs={self._gradient_kwargs})"
+        return (
+            f"TransformJacobianProducts({self._inner_execute}, gradient_transform={self._gradient_transform}, "
+            f"gradient_kwargs={self._gradient_kwargs}, cache_full_jacobian={self._cache_full_jacobian})"
+        )
 
     def __init__(
         self,
         inner_execute: Callable,
         gradient_transform: "qml.gradients.gradient_transform",
         gradient_kwargs: Optional[dict] = None,
+        cache_full_jacobian: bool = False,
     ):
         if logger.isEnabledFor(logging.DEBUG):  # pragma: no cover
             logger.debug(
-                "TransformJacobianProduct being created with (%s, %s, %s)",
+                "TransformJacobianProduct being created with (%s, %s, %s, %s)",
                 inspect.getsource(inner_execute)
                 if (logger.isEnabledFor(qml.logging.TRACE) and inspect.isfunction(inner_execute))
                 else inner_execute,
                 gradient_transform,
                 gradient_kwargs,
+                cache_full_jacobian,
             )
         self._inner_execute = inner_execute
         self._gradient_transform = gradient_transform
         self._gradient_kwargs = gradient_kwargs or {}
+        self._cache_full_jacobian = cache_full_jacobian
+        self._cache = LRUCache(maxsize=10)
 
     def execute_and_compute_jvp(self, tapes: Batch, tangents: Tuple[Tuple[TensorLike]]):
         if logger.isEnabledFor(logging.DEBUG):  # pragma: no cover
             logger.debug("execute_and_compute_jvp called with (%s, %s)", tapes, tangents)
+
         num_result_tapes = len(tapes)
 
+        if self._cache_full_jacobian:
+            jacs = self.compute_jacobian(tapes)
+            jvps = _compute_jvps(jacs, tangents, tapes)
+            return self._inner_execute(tapes), jvps
```

Comment:
so the main change being made here is that if you ever call `compute_jacobian`, it will cache the results, right? it feels strange that the logical flow is entirely different because caching was requested. in other words, a lot of similar code seems to be re-written between here, `compute_jvp` and `compute_jacobian`. I'm fine with copy-pasting code when the situation calls for it, but I'm not sure what that situation is here

Reply:
I'm fleshing out the docstring for the `cache_full_jacobian` keyword to mention that it is specially for the case of autograd with no caching.

Basically, I don't want the `autograd.py` code to have to check "am i taking the innermost recursive derivatives with a gradient transform without caching?" because then `autograd_execute` would need 3 new keyword arguments.

So I'm making it a more general feature that is selectively used in the autograd case.

Reply:
That all makes sense. I get this is a corner case, but from the perspective of `execution.py` the difference is kinda irrelevant. I was just more thinking to go from:

```python
def func(p):
    x = a(p)
    return b(x)

def func_cached(p):
    x = a(p)
    cache[p] = x
    return b(x)
```

to:

```python
def func(p, _cache=False):
    x = a(p):
    if _cache:
        cache[p] = x
    return b(x)
```

does that make sense? in this case, there's a lot of common code being copy-pasted just to make sure we cache intermediate results. I know it's because tapes and stuff are broken up differently at different times, but I'm worried we're just piling onto this split and duplication without a plan to keep it maintainable.